### DISCUSSION TITLE
Deploy to US Gov Cloud Keyvault version

### DISCUSSION THREAD

**Discussion by willdovetx** [polymind-inc/acmebot#543] [Q&A] [answered]
We were successful in getting the non-keyvault version working in our Azure US Gov cloud.  However we keep getting the following error when attempting to setup the keyvault version running.  I am sure this is a step we have missed but don't know where to start with this.  Any suggestions.

`{"name":"GetCertificates_Orchestrator","instanceId":"fe512fe207e94221bfdb5c379389020c","runtimeStatus":"Failed","input":null,"customStatus":null,"output":"Orchestrator function 'GetCertificates_Orchestrator' failed: The activity function 'GetAllCertificates' failed: \"Failed to deserialize exception from TaskActivity: {\"$type\":\"Azure.Identity.AuthenticationFailedException, Azure.Identity\",\"ClassName\":\"Azure.Identity.AuthenticationFailedException\",\"Message\":\"ManagedIdentityCredential authentication failed: The json containing instance metadata does not contain details about the authority in use: login.microsoftonline.us. See https://aka.ms/msal-net-custom-instance-metadata for more details. \\nSee the troubleshooting guide for more information. https://aka.ms/azsdk/net/identity/managedidentitycredential/troubleshoot\",\"Data\":null,\"InnerException\":{\"$type\":\"Microsoft.Identity.Client.MsalClientException, Microsoft.Identity.Client\",\"IsRetryable\":false,\"ErrorCode\":\"invalid-custom-instance-metadata\",\"Message\":\"The json containing instance metadata does not contain details about the authority in use: login.microsoftonline.us. See https://aka.ms/msal-net-custom-instance-metadata for more details. \",\"Data\":{\"$type\":\"System.Collections.ListDictionaryInternal, System.Private.CoreLib\"},\"InnerException\":null,\"HelpLink\":null,\"Source\":\"Microsoft.Identity.Client\",\"HResult\":-2146233088,\"StackTrace\":\"   at Microsoft.Identity.Client.Instance.Discovery.UserMetadataProvider.GetMetadataOrThrow(String environment, ILoggerAdapter logger)\\r\\n   at Microsoft.Identity.Client.Instance.Discovery.InstanceDiscoveryManager.GetMetadataEntryAsync(AuthorityInfo authorityInfo, RequestContext requestContext, Boolean forceValidation)\\r\\n   at Microsoft.Identity.Client.Instance.AuthorityManager.RunInstanceDiscoveryAndValidationAsync()\\r\\n   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.FetchNewAccessTokenAsync(CancellationToken cancellationToken)\\r\\n   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.ExecuteAsync(CancellationToken cancellationToken)\\r\\n   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)\\r\\n   at Microsoft.Identity.Client.ApiConfig.Executors.ConfidentialClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenForClientParameters clientParameters, CancellationToken cancellationToken)\\r\\n   at Azure.Identity.AbstractAcquireTokenParameterBuilderExtensions.ExecuteAsync[T](AbstractAcquireTokenParameterBuilder`1 builder, Boolean async, CancellationToken cancellationToken)\\r\\n   at Azure.Identity.MsalConfidentialClient.AcquireTokenForClientCoreAsync(String[] scopes, String tenantId, Boolean async, CancellationToken cancellationToken)\\r\\n   at Azure.Identity.MsalConfidentialClient.AcquireTokenForClientAsync(String[] scopes, String tenantId, Boolean async, CancellationToken cancellationToken)\\r\\n   at Azure.Identity.ManagedIdentityClient.AuthenticateAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)\\r\\n   at Azure.Identity.ManagedIdentityCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)\"},\"HelpURL\":null,\"StackTraceString\":\"   at Azure.Identity.CredentialDiagnosticScope.FailWrapAndThrow(Exception ex, String additionalMessage)\\r\\n   at Azure.Identity.ManagedIdentityCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)\\r\\n   at Azure.Identity.ManagedIdentityCredential.GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)\\r\\n   at Azure.Identity.DefaultAzureCredential.GetTokenFromSourcesAsync(TokenCredential[] sources, TokenRequestContext requestContext, Boolean async, CancellationToken cancellationToken)\\r\\n   at Azure.Identity.DefaultAzureCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)\\r\\n   at Azure.Identity.CredentialDiagnosticScope.FailWrapAndThrow(Exception ex, String additionalMessage)\\r\\n   at Azure.Identity.DefaultAzureCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)\\r\\n   at Azure.Identity.DefaultAzureCredential.GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)\\r\\n   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.AccessTokenCache.GetHeaderValueFromCredentialAsync(TokenRequestContext context, Boolean async, CancellationToken cancellationToken)\\r\\n   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.AccessTokenCache.GetHeaderValueAsync(HttpMessage message, TokenRequestContext context, Boolean async)\\r\\n   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.AccessTokenCache.GetHeaderValueAsync(HttpMessage message, TokenRequestContext context, Boolean async)\\r\\n   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.AuthenticateAndAuthorizeRequestAsync(HttpMessage message, TokenRequestContext context)\\r\\n   at Azure.Security.KeyVault.ChallengeBasedAuthenticationPolicy.AuthorizeRequestInternal(HttpMessage message, Boolean async)\\r\\n   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)\\r\\n   at Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)\\r\\n   at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)\\r\\n   at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)\\r\\n   at Azure.Core.Pipeline.HttpPipeline.SendRequestAsync(Request request, CancellationToken cancellationToken)\\r\\n   at Azure.Security.KeyVault.KeyVaultPipeline.SendRequestAsync(Request request, CancellationToken cancellationToken)\\r\\n   at Azure.Security.KeyVault.KeyVaultPipeline.GetPageAsync[T](Uri firstPageUri, String nextLink, Func`1 itemFactory, String operationName, CancellationToken cancellationToken)\\r\\n   at Azure.Core.PageResponseEnumerator.FuncAsyncPageable`1.AsPages(String continuationToken, Nullable`1 pageSizeHint)+MoveNext()\\r\\n   at Azure.Core.PageResponseEnumerator.FuncAsyncPageable`1.AsPages(String continuationToken, Nullable`1 pageSizeHint)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()\\r\\n   at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()\\r\\n   at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()\\r\\n   at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()\\r\\n   at KeyVault.Acmebot.Functions.SharedActivity.GetAllCertificates(Object input) in /home/runner/work/keyvault-acmebot/keyvault-acmebot/KeyVault.Acmebot/Functions/SharedActivity.cs:line 87\\r\\n   at KeyVault.Acmebot.Functions.SharedActivity.GetAllCertificates(Object input) in /home/runner/work/keyvault-acmebot/keyvault-acmebot/KeyVault.Acmebot/Functions/SharedActivity.cs:line 87\\r\\n   at Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`2.InvokeAsync(Object instance, Object[] arguments) in C:\\\\projects\\\\azure-webjobs-sdk-rqm4t\\\\src\\\\Microsoft.Azure.WebJobs.Host\\\\Executors\\\\FunctionInvoker.cs:line 52\\r\\n   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.InvokeWithTimeoutAsync(IFunctionInvoker invoker, ParameterHelper parameterHelper, CancellationTokenSource timeoutTokenSource, CancellationTokenSource functionCancellationTokenSource, Boolean throwOnTimeout, TimeSpan timerInterval, IFunctionInstance instance) in C:\\\\projects\\\\azure-webjobs-sdk-rqm4t\\\\src\\\\Microsoft.Azure.WebJobs.Host\\\\Executors\\\\FunctionExecutor.cs:line 581\\r\\n   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithWatchersAsync(IFunctionInstanceEx instance, ParameterHelper parameterHelper, ILogger logger, CancellationTokenSource functionCancellationTokenSource) in C:\\\\projects\\\\azure-webjobs-sdk-rqm4t\\\\src\\\\Microsoft.Azure.WebJobs.Host\\\\Executors\\\\FunctionExecutor.cs:line 527\\r\\n   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithLoggingAsync(IFunctionInstanceEx instance, FunctionStartedMessage message, FunctionInstanceLogEntry instanceLogEntry, ParameterHelper parameterHelper, ILogger logger, CancellationToken cancellationToken) in C:\\\\projects\\\\azure-webjobs-sdk-rqm4t\\\\src\\\\Microsoft.Azure.WebJobs.Host\\\\Executors\\\\FunctionExecutor.cs:line 306\",\"RemoteStackTraceString\":null,\"RemoteStackIndex\":0,\"ExceptionMethod\":null,\"HResult\":-2146233088,\"Source\":\"Azure.Identity\",\"WatsonBuckets\":null}\". See the function execution logs for additional details.","createdTime":"2022-12-14T17:49:05Z","lastUpdatedTime":"2022-12-14T17:49:05Z"}`

**Comment by shibayan** [comment, marked as answer]
It appears to be a bug in the library we are using. We will update the package after the corrected version is released.

https://github.com/Azure/azure-sdk-for-net/issues/32498

**Reply by shibayan** [reply]
Related issue #545

**Reply by willdovetx** [reply]
Thank you

**Comment by willdovetx** [comment]
Shibayan, 

Thanks for the version 4.2.0 workaround.  got past the error I was getting. Now I am having another issue.  I am willing to bet I forgot to give something access, but I cannot find it.

The DNS has been populated for more than 3 months so I am not sure what is going on 

`Exception while executing function: Dns01Precondition The delegated name server is not correct. DNS zone = <changed>, Expected = ns1-08.azuregov-dns.us,ns2-08.azuregov-dns.us,ns3-08.azuregov-dns.us,ns4-08.azuregov-dns.us, Actual =  `

**Reply by willdovetx** [reply]
I had to remove the actual domain so I put `<changed>` in its place

**Reply by shibayan** [reply]
Are you delegating your used domains to Azure DNS? This error will appear if NS records are not properly set.
https://learn.microsoft.com/en-us/azure/dns/dns-delegate-domain-azure-dns

**Reply by willdovetx** [reply]
The DNS for the subdomains were configured as different DNS zones.  I deleted the subdomain DNS zone and deployed as Child record to the primary domain, and it resolved the issue.  I figured it was a user error. Still learning to use this amazing tool.  Thanks
